### ISSUE TITLE
Alpaca2-7b微调后合并模型成功，但是回答的结果相比Alpaca2-7b没有任何变化

### ISSUE THREAD

**Issue by dengxiang1995** [ymcui/Chinese-LLaMA-Alpaca-2#241]
### 提交前必须检查以下项目

- [X] 请确保使用的是仓库最新代码（git pull），一些问题已被解决和修复。
- [X] 我已阅读[项目文档](https://github.com/ymcui/Chinese-LLaMA-Alpaca-2/wiki)和[FAQ章节](https://github.com/ymcui/Chinese-LLaMA-Alpaca-2/wiki/常见问题)并且已在Issue中对问题进行了搜索，没有找到相似问题和解决方案。
- [X] 第三方插件问题：例如[llama.cpp](https://github.com/ggerganov/llama.cpp)、[LangChain](https://github.com/hwchase17/langchain)、[text-generation-webui](https://github.com/oobabooga/text-generation-webui)等，同时建议到对应的项目中查找解决方案。

### 问题类型

模型转换和合并

### 基础模型

Chinese-Alpaca-2 (7B/13B)

### 操作系统

Linux

### 详细描述问题

“”“
**Alpaca2-7b sft微调关键参数**
”“”
lr=1e-4
lora_rank=8
lora_alpha=32
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
modules_to_save="embed_tokens,lm_head"
lora_dropout=0.05

**_pretrained_model=/code/models/chinese-alpaca-2-7b
chinese_tokenizer_path=/code/Chinese-LLaMA-Alpaca-2/scripts/tokenizer
dataset_dir=/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_data/safe_1930_**

“”“
**合并模型参数**
”“”
python merge_llama2_with_chinese_lora_low_mem.py \
    **--base_model /code/models/chinese-alpaca-2-7b \
    --lora_model /code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model \_**
    --output_type huggingface \
--output_dir /code/Chinese-LLaMA-Alpaca-2/scripts/training/merge_output/merge_sft_alpaca2_safe0905

  **其它参数跟源码保持一致的，模型合并后回答结果和合并前一模一样**

### 依赖情况（代码类问题务必提供）

```
# 请在此处粘贴依赖情况（请粘贴在本代码块里）
```
**transformers=4.31.0
peft=0.3.0
torch=2.0.1**

### 运行日志或截图

“”“
”“”
(alpaca2) root@me51babd92804c05bbedc3d4544ef74e-task0-0:/code/Chinese-LLaMA-Alpaca-2/scripts# python merge_llama2_with_chinese_lora_low_mem.py \
    --base_model /code/models/chinese-alpaca-2-7b \
    --lora_model /code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model \
    --output_type huggingface \
    --output_dir /code/Chinese-LLaMA-Alpaca-2/scripts/training/merge_output/merge_sft_alpaca2_safe0905
[2023-09-06 08:58:09,173] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
================================================================================
Base model: /code/models/chinese-alpaca-2-7b
LoRA model: /code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model
Loading /code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model
You are using the legacy behaviour of the <class 'transformers.models.llama.tokenization_llama.LlamaTokenizer'>. This means that tokens that come after special tokens will not be properly handled. We recommend you to read the related pull request available at https://github.com/huggingface/transformers/pull/24565
Loading ckpt pytorch_model-00001-of-00002.bin
Merging...
Saving ckpt pytorch_model-00001-of-00002.bin to /code/Chinese-LLaMA-Alpaca-2/scripts/training/merge_output/merge_sft_alpaca2_safe0905 in HF format...
Loading ckpt pytorch_model-00002-of-00002.bin
Merging...
Saving ckpt pytorch_model-00002-of-00002.bin to /code/Chinese-LLaMA-Alpaca-2/scripts/training/merge_output/merge_sft_alpaca2_safe0905 in HF format...
Saving tokenizer
Saving config.json from /code/models/chinese-alpaca-2-7b
Saving generation_config.json from /code/models/chinese-alpaca-2-7b
Saving pytorch_model.bin.index.json from /code/models/chinese-alpaca-2-7b
Done.
Check output dir: /code/Chinese-LLaMA-Alpaca-2/scripts/training/merge_output/merge_sft_alpaca2_safe0905

@ymcui #

**Comment by airaria** [comment]
可以用sha256sum命令检查一下合并前后模型文件是否发生了变化：
`sha256sum pytorch_model-*bin`

chinese-alpaca-2-7b的sha256的值为：
```
182d552f893db16f7196afdf6ec13f05b43650985c9ce71f41f6d576e978c53f pytorch_model-00001-of-00002.bin 
1f22e3d669ea912c4b0428fb4095f2068eb6192c89406ddd441de87ccec0c1b5 pytorch_model-00002-of-00002.bin 
```

**Comment by dengxiang1995** [comment]
> 可以用sha256sum命令检查一下合并前后模型文件是否发生了变化： `sha256sum pytorch_model-*bin`
> 
> chinese-alpaca-2-7b的sha256的值为：
> 
> ```
> 182d552f893db16f7196afdf6ec13f05b43650985c9ce71f41f6d576e978c53f pytorch_model-00001-of-00002.bin 
> 1f22e3d669ea912c4b0428fb4095f2068eb6192c89406ddd441de87ccec0c1b5 pytorch_model-00002-of-00002.bin 
> ```检查了，sha256的值没有任何变化，我的微调以及合并模型的参数都在上面，能看出来什么问题吗

**Comment by airaria** [comment]
合并后的模型和chinese-alpaca-2-7b没有任何变化吗
ll一下你的`/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model`目录，其中文件大小是否正常？

**Comment by dengxiang1995** [comment]
> 合并后的模型和chinese-alpaca-2-7b没有任何变化吗 ll一下你的`/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model`目录，其中文件大小是否正常？
**(alpaca2) root@me51babd92804c05bbedc3d4544ef74e-task0-0:/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model# ll**
total 848
drwxr-xr-x  2 root root    165 Sep  6 02:17 ./
drwxr-xr-x 10 root root   4096 Sep  6 02:17 ../
-rw-r--r--  1 root root    467 Sep  6 02:17 adapter_config.json
-rw-r--r--  1 root root    443 Sep  6 02:17 adapter_model.bin
-rw-r--r--  1 root root    435 Sep  6 02:17 special_tokens_map.json
-rw-r--r--  1 root root 844403 Sep  6 02:17 tokenizer.model
-rw-r--r--  1 root root    765 Sep  6 02:17 tokenizer_config.json

感谢，这是我alpaca2微调后生成的sft_lora_model里的文件情况，我微调的参数上面有，微调这一块应该没问题吧？

**Comment by airaria** [comment]
> > 合并后的模型和chinese-alpaca-2-7b没有任何变化吗 ll一下你的`/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model`目录，其中文件大小是否正常？
> > **(alpaca2) root@me51babd92804c05bbedc3d4544ef74e-task0-0:/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model# ll**
> > total 848
> > drwxr-xr-x  2 root root    165 Sep  6 02:17 ./
> > drwxr-xr-x 10 root root   4096 Sep  6 02:17 ../
> > -rw-r--r--  1 root root    467 Sep  6 02:17 adapter_config.json
> > -rw-r--r--  1 root root    443 Sep  6 02:17 adapter_model.bin
> > -rw-r--r--  1 root root    435 Sep  6 02:17 special_tokens_map.json
> > -rw-r--r--  1 root root 844403 Sep  6 02:17 tokenizer.model
> > -rw-r--r--  1 root root    765 Sep  6 02:17 tokenizer_config.json
> 
> 感谢，这是我alpaca2微调后生成的sft_lora_model里的文件情况，我微调的参数上面有，微调这一块应该没问题吧？

`adapter_model.bin`的大小不对。你用的是不是deepspeed zero3?

你看一下在上一层目录`/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/`中是不是也有apdater_model.bin？有的话，合并时指定上一层目录，即`/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/`

**Comment by dengxiang1995** [comment]
> > > 合并后的模型和chinese-alpaca-2-7b没有任何变化吗 ll一下你的`/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model`目录，其中文件大小是否正常？
> > > **(alpaca2) root@me51babd92804c05bbedc3d4544ef74e-task0-0:/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/sft_lora_model# ll**
> > > total 848
> > > drwxr-xr-x  2 root root    165 Sep  6 02:17 ./
> > > drwxr-xr-x 10 root root   4096 Sep  6 02:17 ../
> > > -rw-r--r--  1 root root    467 Sep  6 02:17 adapter_config.json
> > > -rw-r--r--  1 root root    443 Sep  6 02:17 adapter_model.bin
> > > -rw-r--r--  1 root root    435 Sep  6 02:17 special_tokens_map.json
> > > -rw-r--r--  1 root root 844403 Sep  6 02:17 tokenizer.model
> > > -rw-r--r--  1 root root    765 Sep  6 02:17 tokenizer_config.json
> > 
> > 
> > 感谢，这是我alpaca2微调后生成的sft_lora_model里的文件情况，我微调的参数上面有，微调这一块应该没问题吧？
> 
> `adapter_model.bin`的大小不对。你用的是不是deepspeed zero3?
> 
> 你看一下在上一层目录`/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8/`中是不是也有apdater_model.bin？有的话，合并时指定上一层目录，即`/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/`

用的ds_zero2,/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/这层目录没有adapter_model.bin这个文件

**Comment by dengxiang1995** [comment]
**这是sft微调全部参数**
lr=1e-4
lora_rank=8
lora_alpha=32
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
modules_to_save="embed_tokens,lm_head"
lora_dropout=0.05

pretrained_model=/code/models/chinese-alpaca-2-7b
chinese_tokenizer_path=/code/Chinese-LLaMA-Alpaca-2/scripts/tokenizer
dataset_dir=/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_data/safe_1930
# data_cache=/code/Chinese-LLaMA-Alpaca-2/scripts/training/cache/pt_sft_safe_cache_0904_epoch10
per_device_train_batch_size=1
per_device_eval_batch_size=1
gradient_accumulation_steps=2
max_seq_length=512
output_dir=/code/Chinese-LLaMA-Alpaca-2/scripts/training/sft_output/sft_safe_0905_epoch8
validation_file=/code/Chinese-LLaMA-Alpaca-2/scripts/training/verify_data/alpaca2_safe_verify_400.json

deepspeed_config_file=ds_zero2_no_offload.json

torchrun --nnodes 1 --nproc_per_node 1 run_clm_sft_with_peft.py \
    --deepspeed ${deepspeed_config_file} \
    --model_name_or_path ${pretrained_model} \
    --tokenizer_name_or_path ${chinese_tokenizer_path} \
    --dataset_dir ${dataset_dir} \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --per_device_eval_batch_size ${per_device_eval_batch_size} \
    --do_train \
    --do_eval \
    --seed $RANDOM \
    --fp16 \
    --num_train_epochs 8 \
    --lr_scheduler_type cosine \
    --learning_rate ${lr} \
    --warmup_ratio 0.03 \
    --weight_decay 0 \
    --logging_strategy steps \
    --logging_steps 10 \
    --save_strategy steps \
    --save_total_limit 20 \
    --evaluation_strategy steps \
    --eval_steps 100 \
    --save_steps 1000 \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --preprocessing_num_workers 8 \
    --max_seq_length ${max_seq_length} \
    --output_dir ${output_dir} \
    --overwrite_output_dir \
    --ddp_timeout 30000 \
    --logging_first_step True \
    --lora_rank ${lora_rank} \
    --lora_alpha ${lora_alpha} \
    --trainable ${lora_trainable} \
    --lora_dropout ${lora_dropout} \
    --torch_dtype float16 \
    --validation_file ${validation_file} \
    --modules_to_save ${modules_to_save}

**Comment by airaria** [comment]
peft版本和我们推荐的似乎并不一致。
安装指定的peft版本，拉取并用最新的训练代码试试

**Comment by dengxiang1995** [comment]
> peft版本和我们推荐的似乎并不一致。 安装指定的peft版本，拉取并用最新的训练代码试试

要求的peft版本为0.3.0.dev0  这是官方指定的，一样的啊。。

**Comment by dengxiang1995** [comment]
> > peft版本和我们推荐的似乎并不一致。 安装指定的peft版本，拉取并用最新的训练代码试试
> 
> 要求的peft版本为0.3.0.dev0 这是官方指定的，一样的啊。。
requirements.txt里是这个：https://github.com/huggingface/peft.git@13e53fc 但是这个打不开，0.3.0和0.3.0.dev0不一样吗。。

**Comment by airaria** [comment]
不一样。
最新的代码里我们在`scripts/training/`放了一份peft，应该不用额外安装了，所以用最新代码试试

**Comment by dengxiang1995** [comment]
> 不一样。 最新的代码里我们在`scripts/training/`放了一份peft，应该不用额外安装了，所以用最新代码试试

还真有，谢谢，我试试，这个后续是否可以在常见问题里补充，要不然真不知道，peft版本不对是可能微调失败的原因是吧？

**Comment by airaria** [comment]
另外，指定版本的peft我安装没有问题：
```
pip install git+https://github.com/huggingface/peft.git@13e53fc
```

peft的版本之间变动比较大，版本不对的话可能导致各种问题，所以建议使用指定版本。
